File: Trabalho flutter 2/flutter_application_2/lib/models/transaction.dart
/// Modelo Transaction (Transação Financeira)
///
/// Migrado de: br.edu.ifsuldeminas.mch.controlefinanceiro.model.Transaction (Java)
///
/// **Melhorias:**
/// - Enum para tipo de transação (type safety)
/// - DateTime nativo em vez de String
/// - Validações no construtor
class Transaction {
  final int? id;
  final double amount;
  final String description;
  final TransactionType type;
  final bool isTaxExempt;
  final DateTime date;

  Transaction({
    this.id,
    required this.amount,
    required this.description,
    required this.type,
    this.isTaxExempt = false,
    DateTime? date,
  }) : date = date ?? DateTime.now();

  /// Converte para Map (SQLite)
  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'amount': amount,
      'description': description,
      'type': type.name, // Salva como String: 'income' ou 'expense'
      'isTaxExempt': isTaxExempt ? 1 : 0,
      'date': date.toIso8601String(),
    };
  }

  /// Cria Transaction a partir de Map
  factory Transaction.fromMap(Map<String, dynamic> map) {
    return Transaction(
      id: map['id'] as int?,
      amount: map['amount'] as double,
      description: map['description'] as String,
      type: TransactionType.values.firstWhere(
        (e) => e.name == map['type'],
        orElse: () => TransactionType.expense,
      ),
      isTaxExempt: map['isTaxExempt'] == 1,
      date: DateTime.parse(map['date'] as String),
    );
  }

  Transaction copyWith({
    int? id,
    double? amount,
    String? description,
    TransactionType? type,
    bool? isTaxExempt,
    DateTime? date,
  }) {
    return Transaction(
      id: id ?? this.id,
      amount: amount ?? this.amount,
      description: description ?? this.description,
      type: type ?? this.type,
      isTaxExempt: isTaxExempt ?? this.isTaxExempt,
      date: date ?? this.date,
    );
  }

  @override
  String toString() {
    return 'R\$ ${amount.toStringAsFixed(2)} - $description\n'
        '${isTaxExempt ? "Isenta" : "Não Isenta"} - ${type.label}';
  }
}

/// Enum para tipo de transação
///
/// **Por que Enum em vez de String?**
/// - Type safety: impossível passar valor inválido
/// - Autocomplete no IDE
/// - Refatoração segura
enum TransactionType {
  income('Receita'),
  expense('Despesa');

  final String label;
  const TransactionType(this.label);
}

File: Trabalho flutter 2/flutter_application_2/ControleFinanceiro-main/app/src/main/java/br/edu/ifsuldeminas/mch/controlefinanceiro/model/Transaction.java
package br.edu.ifsuldeminas.mch.controlefinanceiro.model;

import androidx.annotation.NonNull;

import java.io.Serializable;

public class Transaction implements Serializable {

    private String id;
    private Double amount;
    private String description;
    private String type;
    private boolean isTaxExempt;


    public Transaction() {
    }

    public Transaction(String id, Double amount, String description, String type, Boolean isTaxExempt) {
        this.id = id;
        this.amount = amount;
        this.description = description;
        this.type = type;
        this.isTaxExempt = isTaxExempt;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Double getAmount() {
        return amount;
    }

    public void setAmount(Double amount) {
        this.amount = amount;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public boolean isTaxExempt() {
        return isTaxExempt;
    }

    public void setTaxExempt(boolean taxExempt) {
        isTaxExempt = taxExempt;
    }

    @NonNull
    @Override
    public String toString() {
        return
                "R$" + amount +
                        " - " + description + "\n" +
                        (isTaxExempt ? "Isenta" : "Não Isenta") + " - " + type;
    }

}
